### PR TITLE
scheduler: Make fenzo bin packing behavior configurable & pluggable.

### PR DESCRIPTION
**Pull request by mforsyth**

### CODE REVIEW

File: scheduler/src/cook/mesos/scheduler.clj [hunk 1]
```diff
@@ -43,12 +43,14 @@
             [plumbing.core :refer (map-vals)]
             [swiss.arrows :refer :all]
             [clojure.core.cache :as cache]
-            [cook.mesos.reason :as reason])
+            [cook.mesos.reason :as reason]
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Lexicographic ordering please

File: scheduler/src/cook/mesos/scheduler.clj [hunk 2]
```diff
@@ -1161,13 +1163,42 @@
...
+(defn config-string->fitness-calculator
+  "Given a string specified in the configuration, attempt to resolve it
+  to and return an instance of com.netflix.fenzo.VMTaskFitnessCalculator.
+  The config string can either be a reference to a clojure symbol, or to a
+  static member of a java class (for example, one of the fitness calculators
+  that ship with Fenzo).  An exception will be thrown if a VMTaskFitnessCalculator
+  can't be found using either method."
+  [config-string]
+  (try
+    ^VMTaskFitnessCalculator (-> config-string symbol resolve deref)
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: Should `VMTaskFitnessCalculator` be placed in the function signature instead?

File: scheduler/src/cook/mesos/scheduler.clj [hunk 3]
```diff
@@ -1161,13 +1163,42 @@
 (meters/defmeter [cook-mesos scheduler mesos-error])
 (meters/defmeter [cook-mesos scheduler offer-chan-full-error])
 
+(def dummy-fitness-calculator
```
Comment (shamsimam) [inline scheduler/src/cook/mesos/scheduler.clj]: If this is used only for testing purposes, we should it move it to a class in the test package.

File: scheduler/src/cook/components.clj [hunk 4]
```diff
@@ -356,6 +357,8 @@
                                               fenzo-floor-iterations-before-warn)
      :fenzo-floor-iterations-before-reset (fnk [[:config [:scheduler {fenzo-floor-iterations-before-reset 1000}]]]
                                                fenzo-floor-iterations-before-reset)
+     :fenzo-fitness-calculator (fnk [[:config [:scheduler {fenzo-fitness-calculator "com.netflix.fenzo.plugins.BinPackingFitnessCalculators/cpuMemBinPacker"}]]]
```
Comment (shamsimam) [inline scheduler/src/cook/components.clj]: `com.netflix.fenzo.plugins.BinPackingFitnessCalculators/cpuMemBinPacker` is duplicated in multiple locations, maybe it should be extracted into a constant?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 5]
```diff
@@ -763,5 +764,23 @@
...
+                    "cook.mesos.scheduler/dummy-fitness-calculator"))))
+  (testing "java class on classpath"
+    (is (instance? VMTaskFitnessCalculator
+                   (sched/config-string->fitness-calculator
+                    "com.netflix.fenzo.plugins.BinPackingFitnessCalculators/cpuMemBinPacker"))))
+
+  (testing "bad input"
+    (is (thrown? Exception (sched/config-string->fitness-calculator "not-a-valid-anything"))))
+
+  (testing "something other than a VMTaskFitnessCalculator"
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Can we include more specific Exception or message regex in the tests?

File: scheduler/test/cook/test/mesos/scheduler.clj [hunk 6]
```diff
@@ -763,5 +764,23 @@
...
+  (testing "clojure symbol"
+    (is (instance? VMTaskFitnessCalculator
+                   (sched/config-string->fitness-calculator
+                    "cook.mesos.scheduler/dummy-fitness-calculator"))))
+  (testing "java class on classpath"
+    (is (instance? VMTaskFitnessCalculator
+                   (sched/config-string->fitness-calculator
+                    "com.netflix.fenzo.plugins.BinPackingFitnessCalculators/cpuMemBinPacker"))))
+
+  (testing "bad input"
```
Comment (shamsimam) [inline scheduler/test/cook/test/mesos/scheduler.clj]: Can we include more specific Exception or message regex in the tests?

**Comment by mforsyth** [comment]
@shamsimam  Comments incorporated... I actually found out that type hints aren't enough to ensure that a var implements a specific interface in clojure.  You need to manually check if you want an exception to be thrown in that situation.  I also rebased off of new master to eliminate the conflicts.

**Comment by wyegelwel** [comment]
This LGTM. @shamsimam, any additional comments?

**Comment by shamsimam** [comment]
LGTM.

**Pull request merged**